### ISSUE TITLE
destruct in pattern match seems to break exhaustiveness checking 

### ISSUE THREAD

**Issue by hjwp** [python/mypy#13782]
**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.10&gist=96b3ccda6af55153c47093c2228a5370

```python
from dataclasses import dataclass
from typing import List, Literal, NoReturn

def assert_never(value: NoReturn) -> NoReturn:
    assert False, f'This code should never be reached, got: {value}'

@dataclass
class Cat:
    type: str

@dataclass
class Dog:
    good: Literal[True]


def assess_pets(pets: List[Cat | Dog]) -> str:
    for c in pets:
        match c:
            case Cat(type=t) if t == "tabby":
                return "delightful"
            case Cat():
                return "fabulous"
            case Dog():
                return "good boy!"
        assert_never(c)
```

**Expected Behavior**

This code should show no errors

**Actual Behavior**

```
main.py:25: error: Argument 1 to "assert_never" has incompatible type "Cat"; expected "NoReturn"
```

it seems that the `Cat(type=t)` destructuring is confusing mypy, and it no longer recognises the `case Cat()` that follows as being an exhaustive catch-all. 

if you comment out the `type=t` option then the error disappears.

(mypy also complains about a missing return statement which i would argue is also an error?)

```
main.py:16: error: Missing return statement
```

the match is exhaustive, there is no missing return statement.

**Your Environment**

* Python 3.10.6
* mypy 0.981 (compiled: yes)
* no config, no cli args.

**Comment by erictraut** [comment]
I think mypy is correct in this case. 

You're using `assert_never` incorrectly in your example. It should be used to determine whether the argument (in your case, the expression `c`) evaluates to the type `Never`. But in your example, the `assert_never(c)` statement is unreachable, so the narrowed type of `c` is indeterminate at that point.

Perhaps you intended the following instead?

```python
def assess_pets(pets: List[Cat | Dog]) -> str:
    for c in pets:
        match c:
            case Cat(type=t) if t == "tabby":
                return "delightful"
            case Cat():
                return "fabulous"
            case Dog():
                return "good boy!"
            case x:
                assert_never(x)
```

Mypy is also correct in indicating that there is a missing return statement because `pets` could be a list of length zero, which means control would fall through the `for` loop, and the `match` statement would never be executed.

**Comment by hjwp** [comment]
thanks!  I think you're right about the "missing return statement" thing, I get that now.

I"m still worried about the `assert_never()`.  How do you explain that the type error disappears if you comment out the first `match Cat(type=t)` ?

**Comment by erictraut** [comment]
I suspect that the error disappears because the `assert_never(c)` statement is proven unreachable, so it not analyzed.

The `assert_never(c)` statement should be provably unreachable with or without the first match statement present, so I agree that something looks amiss here.

**Comment by hjwp** [comment]
should we reopen the bug then?
